components/content_settings/renderer/brave_content_settings_agent_impl.cc

```diff
@@ -135,6 +144,62 @@ void BraveContentSettingsAgentImpl::DidNotAllowScript() {
   ContentSettingsAgentImpl::DidNotAllowScript();
 }
 
+bool BraveContentSettingsAgentImpl::UseEphemeralStorageSync(
+    StorageType storage_type) {
+  if (!base::FeatureList::IsEnabled(net::features::kBraveEphemeralStorage))
+    return false;
+
+  if (storage_type != StorageType::kLocalStorage &&
+      storage_type != StorageType::kSessionStorage)
+    return false;
+
+  blink::WebLocalFrame* frame = render_frame()->GetWebFrame();
+
+  if (!frame || IsFrameWithOpaqueOrigin(frame))
+    return false;
+
+  return  // block 3p
+      !ContentSettingsAgentImpl::AllowStorageAccessSync(storage_type) &&
+      // allow 1p
+      AllowStorageAccessForMainFrameSync(storage_type);
+}
+
+bool BraveContentSettingsAgentImpl::AllowStorageAccessSync(
+    StorageType storage_type) {
+  bool result = ContentSettingsAgentImpl::AllowStorageAccessSync(storage_type);
+
+  if (result || UseEphemeralStorageSync(storage_type))
+    return true;
+
+  return false;
+}
+
+bool BraveContentSettingsAgentImpl::AllowStorageAccessForMainFrameSync(
+    StorageType storage_type) {
+  blink::WebLocalFrame* frame = render_frame()->GetWebFrame();
+
+  if (IsFrameWithOpaqueOrigin(frame))
+    return false;
+
+  bool result = false;
+
+  StoragePermissionsKey key(url::Origin(frame->GetSecurityOrigin()),
+                            storage_type);
+  const auto permissions = cached_storage_permissions_.find(key);
+  if (permissions != cached_storage_permissions_.end())
+    return permissions->second;
+
+  // check for block all by looking at top domain only
+  GetContentSettingsManager().AllowStorageAccess(
+      routing_id(), ConvertToMojoStorageType(storage_type),
+      frame->GetDocument().TopFrameOrigin(),
+      url::Origin(frame->GetDocument().TopFrameOrigin()).GetURL(),
+      frame->GetDocument().TopFrameOrigin(), &result);
+  cached_storage_permissions_[key] = result;
```

Comment:
do we need the cache at all? Growing it indefinitely looks pretty scary

Reply:
this is copied from ContentSettingsAgentImpl and it's per-frame so it doesn't grow indefinitely